n3fit/src/n3fit/model_gen.py

```diff
@@ -780,44 +680,120 @@ def compute_unnormalized_pdf(x, postfix=""):
     regularizer_args: dict,
     last_layer_nodes: int,
-    name: str,
 ) -> MetaModel:
     """
     Create the part of the model that contains all of the actual neural network
-    layers.
+    layers, for each replica.
+
+    Parameters
+    ----------
+        layer_type: str
+            Type of layer to use. Can be "dense" or "dense_per_flavour".
+        nodes_in: int
+            Number of nodes in the input layer.
+        nodes: List[int]
+            Number of nodes in each hidden layer.
+        activations: List[str]
+            Activation function to use in each hidden layer.
+        initializer_name: str
+            Name of the initializer to use.
+        replica_seeds: List[int]
+            List of seeds to use for each replica.
+        dropout: float
+            Dropout rate to use (if 0, no dropout is used).
+        regularizer: str
+            Name of the regularizer to use.
+        regularizer_args: dict
+            Arguments to pass to the regularizer.
+        last_layer_nodes: int
+            Number of nodes in the last layer.
+
+    Returns
+    -------
+        nn_replicas: List[MetaModel]
+            List of MetaModel objects, one for each replica.
     """
-    common_args = {
-        'nodes_in': input_dimensions,
-        'nodes': nodes,
-        'activations': activations,
-        'initializer_name': initializer_name,
-        'seed': replica_seed,
-    }
-    if layer_type == "dense":
+    nodes_list = list(nodes)  # so we can modify it
+    x = Input(shape=(None, nodes_in), batch_size=1, name='xgrids_processed')
+
+    custom_args = {}
+    if layer_type == "dense_per_flavour":
+        # set the arguments that will define the layer
+        # but careful, the last layer must be nodes = 1
+        # TODO the mismatch is due to the fact that basis_size
+        # is set to the number of nodes of the last layer when it should
+        # come from the runcard
+        nodes_list[-1] = 1
+        basis_size = last_layer_nodes
+        custom_args['basis_size'] = basis_size
+
+        def initializer_generator(initializer_name, seed, i_layer):
+            seed += i_layer * basis_size
+            initializers = [
+                MetaLayer.select_initializer(initializer_name, seed=seed + b)
+                for b in range(basis_size)
+            ]
+            return initializers
+
+    elif layer_type == "dense":
         reg = regularizer_selector(regularizer, **regularizer_args)
-        list_of_pdf_layers = generate_dense_network(
-            **common_args, dropout_rate=dropout, regularizer=reg
-        )
-    elif layer_type == "dense_per_flavour":
-        list_of_pdf_layers = generate_dense_per_flavour_network(
-            **common_args, basis_size=last_layer_nodes
-        )
+        custom_args['regularizer'] = reg
+
+        def initializer_generator(initializer_name, seed, i_layer):
+            seed += i_layer
+            return MetaLayer.select_initializer(initializer_name, seed=seed)
+
```

Comment:
Maybe add a `raise ValueError` here if `layer_type` is not either `dense` or `dense_per_flavour`. This would at least remove the warning that `initializer_generator` below is possible unbound.

Ideally, this would have been checked in [n3fit.checks](initializer_generator) (?).

```suggestion
    else:
        raise ValueError(f"{layer_type} not allowed. Can only be 'dense' or 'dense_per_flavour'")
```

Reply:
You mean a warning you get in your IDE, from a linter or something? 
I can add a check in checks, but I think that won't remove this?

Reply:
That will remove the warning (but this is a very minor part of it). The main reason it to have it crash with informative error message if it is not one of the two.

Reply:
I would vote for a `check_layer_accepted_type` or something in `n3fit/checks.py` rather than having the check here, otherwise it will fail very late (after having loaded all fktables!)

Reply:
Ok, done!

Reply:
I called it from `wrapper_check_NN` , which is already in `n3fit_checks_provider.py`, and it seems to work fine for me, is that not enough?

Reply:
Oh, missed that line! Yes, that's enough.

I'll have a second look later today if I can and merge it. Thanks!

Reply:
> I called it from `wrapper_check_NN` , which is already in `n3fit_checks_provider.py`, and it seems to work fine for me, is that not enough?

Yes, this would be the ideal solution indeed.

Reply:
Ok great, thanks :)